src/wasm/wasm-type.cpp

```diff
@@ -2276,191 +2271,174 @@ void TypeBuilder::createRecGroup(size_t index, size_t length) {
+          i, TypeBuilder::ErrorReason::ForwardSupertypeReference}};
+      }
+      // The supertype must have a valid structure.
+      if (!isValidSupertype(*info, *super)) {
+        return {
+          TypeBuilder::Error{i, TypeBuilder::ErrorReason::InvalidSupertype}};
+      }
     }
+    seenTypes.insert(asHeapType(info));
   }
 
-  // Remove replaced types from newInfos.
-  bool needsConsolidation = false;
-  for (auto& info : newInfos) {
-    HeapType oldType = asHeapType(info);
-    if (auto it = replacements.find(oldType); it != replacements.end()) {
-      auto& replacement = it->second;
-      if (auto* newInfo = replacement.getHeapTypeInfo()) {
-        // Replace the old info with the new info in `newInfos`, replacing the
-        // moved info in `replacement` with the corresponding new HeapType so we
-        // can still perform the correct replacement in the next step.
-        HeapType newType = asHeapType(*newInfo);
-        info = std::move(*newInfo);
-        replacement = {newType};
-      } else {
-        // The old info is replaced, but there is no new Info to replace it, so
-        // just delete the old info. We will remove the holes in `newInfos`
-        // afterwards.
-        info = nullptr;
-        needsConsolidation = true;
+  // Check that children are either already canonical (i.e. defined in a
+  // previous rec group) or are defined in this current rec group.
+  for (size_t i = 0; i < typeInfos.size(); ++i) {
+    auto type = asHeapType(typeInfos[i]);
+    for (auto child : type.getHeapTypeChildren()) {
+      if (isTemp(child) && !seenTypes.count(child)) {
+        return {TypeBuilder::Error{
+          i, TypeBuilder::ErrorReason::ForwardChildReference}};
       }
     }
   }
-  if (needsConsolidation) {
-    newInfos.erase(std::remove(newInfos.begin(), newInfos.end(), nullptr),
-                   newInfos.end());
-  }
-}
 
-void CanonicalizationState::updateUses(ReplacementMap& replacements,
-                                       std::unique_ptr<HeapTypeInfo>& info) {
-  if (replacements.empty()) {
-    return;
-  }
-  // Replace all old types reachable from `info`.
-  struct ChildUpdater : HeapTypeChildWalker<ChildUpdater> {
-    ReplacementMap& replacements;
-    ChildUpdater(ReplacementMap& replacements) : replacements(replacements) {}
-    void noteChild(HeapType* child) {
-      if (auto it = replacements.find(*child); it != replacements.end()) {
-        *child = it->second.getAsHeapType();
-      }
+  // The rec group is valid, so we can try to move the group into the global rec
+  // group store. If the returned canonical group is not the same as the input
+  // group, then there is already a canonical version of this rec group. Lock
+  // the global rec group store here to avoid leaking temporary types to other
+  // threads that may be trying to build an identical group.
+  auto group = asHeapType(typeInfos[0]).getRecGroup();
+  std::lock_guard<std::mutex> lock(globalRecGroupStore.mutex);
```

Comment:
Yes, because we wouldn't want to leak any temporary, non-canonicalized Types to the rest of the program. If we released this lock early, then a parallel call to `TypeBuilder::build()` could try to canonicalize an identical rec group, get back the rec group we are currently building, and end up returning references to types we have not finished canonicalizing yet.

Reply:
Got it, makes sense.